Vm cleanup

bjourne opened the pull request:
Here is a small pr to make the cpp code cleaner. I made a macro `FACTOR_FOR_EACH` to make iteration nicer. I also added some functions to io.cpp so that if you run `factor -i=/tmp` it doesn't crash.

mrjbq7 commented on vm/utilities.hpp:
```diff
@@ -1,5 +1,11 @@
 
+// Poor mans range-based for loops.
+#define FACTOR_FOR_EACH(iterable)                           \
+  for (typeof((iterable).begin()) iter = (iterable).begin();    \
```
Does this call ``begin()`` twice?

mrjbq7 commented on vm/utilities.hpp:
```diff
@@ -1,5 +1,11 @@
+// Poor mans range-based for loops.
+#define FACTOR_FOR_EACH(iterable)                           \
+  for (typeof((iterable).begin()) iter = (iterable).begin();    \
+       iter != (iterable).end();                                \
```
And does this call ``end()`` on each loop?

mrjbq7 commented on vm/image.cpp:
```diff
@@ -253,7 +252,7 @@ void factor_vm::load_image(vm_parameters* p) {
   load_data_heap(file, &h, p);
   load_code_heap(file, &h, p);
 
-  safe_fclose(file);
```
Why change this one?  To hide ``io-error`` if close fails?

erg replied:
``safe_fclose`` is wrong anyway -- you're not allowed to retry on ``EINTR``
on ``flclose`` or ``close``.

On Wed, Jun 3, 2015 at 11:24 AM, John Benediktsson <notifications@github.com
> wrote:

> In vm/image.cpp
> <https://github.com/slavapestov/factor/pull/1334#discussion_r31653622>:
>
> > @@ -253,7 +252,7 @@ void factor_vm::load_image(vm_parameters* p) {
> >    load_data_heap(file, &h, p);
> >    load_code_heap(file, &h, p);
> >
> > -  safe_fclose(file);
>
> Why change this one? To hide io-error if close fails?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/slavapestov/factor/pull/1334/files#r31653622>.
>

bjourne replied:
It's replaced with `raw_fclose`. Now if it fails, it will not segfault Factor.

mrjbq7 commented on vm/io.cpp:
```diff
@@ -13,6 +13,39 @@ The Factor library provides platform-specific code for Unix and Windows
 normal operation. */
 
+int raw_fread(void* ptr, size_t size, size_t nitems, FILE* stream) {
+  size_t items_read = 0;
```
Seems weird to return a ``size_t`` cast to an ``int``.

mrjbq7 replied:
Maybe an ``ssize_t`` and overflow detection?  I dunno.

mrjbq7 commented on vm/io.cpp:
```diff
@@ -13,6 +13,39 @@ The Factor library provides platform-specific code for Unix and Windows
+  do {
+    size_t ret = fread((void*)((int*)ptr + items_read * size), size,
+                       nitems - items_read, stream);
+    if (ret == 0) {
```
The man pages say:

| If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

So perhaps you need to check ``if (ret <= expected)``.

bjourne replied:
Ok, I'll look into it closer. The idea was to return `-1` on error, therefore the size_t -> int change.

bjourne commented:
Updated the pr. The `FACTOR_FOR_EACH` is fixed so now end() is only called once. I also changed the raw_fread/safe_fread combo back to returning a size_t. It returns 0 on error.

mrjbq7 commented:
Looks pretty good to me, @erg?

erg commented:
Looks good.

On Thu, Jun 4, 2015 at 11:05 PM, John Benediktsson <notifications@github.com
> wrote:

> Looks pretty good to me, @erg <https://github.com/erg>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/slavapestov/factor/pull/1334#issuecomment-109172366>.
>

mrjbq7 commented:
Merged, thanks!!
